crates/base-db/src/input.rs

```diff
@@ -351,26 +411,43 @@ impl Env {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum DependencyKind {
+    Normal,
+    Dev,
+    Build,
+}
```

Comment:
Please don't take this too seriously, as I have retreated to my Zig ivory tower in Lisbon some years ago and am mostly out of touch with the current goings of rust-analyzer, but I would say this is a significant architectural bug. 

By design, this layer of rust-analyzer knows nothing about cargo specific concepts. dev-vs-build-vs normal is 100% Cargo concept. `rustc` knows nothing about these words. As such, _any_ special handling of these concepts should happen in `workspace.rs`, not here. 

The motivation for this design is two-fold:

- practically, Cargo is _a_ build system, there might be others. In the core analysis parts, we shouldn't be prioritizing Cargo over other (current _and future_) build systems
- theoretically, it is important that the internal objects in rust-analyzer's semantic model reflect the physical reality of rustc. In that physical reality, cargo packages do not exists (and `-dev` as a concept applies to a package), there are only units of compilation.

cc @Veykril

Reply:
> Please don't take this too seriously, 

Not at all, you are completely right, this seems incorrect. This was a mistake on my part (I proposed this change somewhere I believe, might've been in DMs).

Reply:
In order to check for equality we needed to filter out dev dependencies and thus we ended up with the current state of things.  The alternatives were 

1. To filter them in `project_model` : I am looking into this since @matklad 's first comment here, I do not think it is going to be easy.
2. Adding `project-model` as a dependency of `base-db` which is ofc super wrong to do. ( As a matter of fact we already do have a `DependencyKind` under `project-model` named `DepKind` )

Reply:
But let me say that I also agree with matklad's concerns so I will look for a way to rectify this.